python/ray/data/datasource/parquet_datasource.py

```diff
@@ -56,6 +57,70 @@ def _deregister_parquet_file_fragment_serialization():
     ray.util.deregister_serializer(ParquetFileFragment)
 
 
+# easier to test when isolated from retry logic
+def _deserialize_pieces(
+    serialized_pieces: str,
+) -> List["pyarrow._dataset.ParquetFileFragment"]:
+    from ray import cloudpickle
+
+    pieces: List["pyarrow._dataset.ParquetFileFragment"] = cloudpickle.loads(
+        serialized_pieces
+    )
+    return pieces
+
+
+# this retry helps when the upstream datasource is not able to handle
+# overloaded read request or failed with some retriable failures.
+# For example when reading data from HA hdfs service, hdfs might
+# lose connection for some unknown reason expecially when
+# simutaneously running many hyper parameter tuning jobs
+# with ray.data parallelism setting at high value like the default 200
+# Such connection failure can be restored with some waiting and retry.
+def _deserialize_pieces_with_retry(
+    serialized_pieces: str,
+) -> List["pyarrow._dataset.ParquetFileFragment"]:
+    min_interval = 0
+    final_exception = None
+    for i in range(FILE_READING_RETRY):
+        try:
+            return _deserialize_pieces(serialized_pieces)
+        except Exception as e:
+            import traceback
+            import time
+            import random
+
+            tb_str = traceback.format_exception(
+                etype=type(e), value=e, tb=e.__traceback__
+            )
+            err_msg_str = f"{type(e)}:{str(e)}"
+            retry_timing = "" if i == 7 else (f" Retry after {min_interval} sec. ")
+            log_only_show_in_1st_retry = (
+                ""
+                if i
+                else (
+                    f"If earlier read attempt threw certain Exception"
+                    f", It may or may not be an issue depends on these retries "
+                    f"succeed or not. serialized_pieces:{serialized_pieces}"
+                )
+            )
```

Comment:
I actually don't think this is an important message to give to the user, since you already say in `retry_timing` that you are retrying the read.

Reply:
I probably will keep this one.  Because when real retry situation happens, hdfs will throw a very elaborated java exception about connection issue.  This exception can not be captured by python try: except: clause.  Our customer will 100% ask us what was that about.  Therefore, I add this line here to (hopefully) remind them if later retry succeeds, they can ignore the exception threw by java.